File: zhanglianzhuang/day5/demo05/src/main/java/com/example/demo05/controller/WebSocketController.java
package com.example.demo05.controller;

import org.springframework.stereotype.Component;
import javax.websocket.*;
import javax.websocket.server.PathParam;
import javax.websocket.server.ServerEndpoint;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Component
@ServerEndpoint("/websocket/{pageCode}")
public class WebSocketController {
	private static final String loggeerName=WebSocketController.class.getName();

	public static Map<String,List<Session>> electricSocketmap = new ConcurrentHashMap<String, List<Session>>();
	//建立连接对象
	@OnOpen
	public void onOpen(@PathParam("pageCode") String pageCode, Session session){
		List<Session> sessions=electricSocketmap.get(pageCode);
		if(sessions==null){
			List<Session> sessionList=new ArrayList<>();
			sessionList.add(session);
			electricSocketmap.put(pageCode,sessionList);
		}
		else {
			sessions.add(session);
		}
	}

	//收到调用的信息
	@OnMessage
	public void onMessage(String message,Session session){
		System.out.println("接受到的信息:"+message);
		try {
			session.getBasicRemote().sendText("我推送的信息是:"+message);
			System.out.println("我推送的信息是:"+message);
		}catch (IOException e){
			System.out.println(e.getMessage());
		}
	}
	@OnError
	public void onError(Session session,Throwable throwable){
		System.out.println("发生错误");
	}
	@OnClose
	public void onClose(@PathParam("pageCode")String pageCode,Session session){
		if (electricSocketmap.containsKey(pageCode)){
			electricSocketmap.get(pageCode).remove(session);

		}
	}
}

File: caiguicheng/day05/websocket/controller/WebSocketController.kt
package com.example.demo05.controller

import org.springframework.stereotype.Component
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap
import javax.websocket.*
import javax.websocket.server.PathParam
import javax.websocket.server.ServerEndpoint

@ServerEndpoint("/websocket/{pageCode}")
@Component
open class WebSocketController {

    private val loggerName = WebSocketController::class.java.name

    var electricScoketmap: MutableMap<String, MutableList<Session>> = ConcurrentHashMap()

    @OnOpen
    open fun onOpen(@PathParam("pageCode") pageCode: String, session: Session) {
        val sessions: MutableList<Session>? = electricScoketmap[pageCode]
        if (sessions == null) {
            val sessionList = ArrayList<Session>()
            sessionList.add(session)
            electricScoketmap[pageCode] = sessionList
        } else {
            electricScoketmap[pageCode]?.add(session)
        }
    }

    //收到调用信息
    @OnMessage
    fun onMessage(message: String, session: Session) {
        println("发送的内容:$message")
//        println("session:$session")
//        val mutableList = electricScoketmap["1"]
//        mutableList?.let {
//            println(it)
//        }
        //session.basicRemote.sendText("我推送的信息是:$message")
        try {
            session.basicRemote.sendText("我发送的内容是:${message}")
        } catch (e: IOException) {
            println(e.message)
        }
    }

    @OnError
    fun onError(session: Session, throwable: Throwable) {
        println("发生错误")
    }

    @OnClose
    fun onClose(@PathParam("pageCode") pageCode: String, session: Session) {
        if (electricScoketmap.containsKey(pageCode)) {
            electricScoketmap[pageCode]?.remove(session)
        }
        //println("close")
    }
}